src/vm_cpu.rs

```diff
@@ -499,22 +696,16 @@ impl VmCpu {
             hstatus.modify(hstatus::vtw.val(1));
         }
         state.guest_regs.hstatus = hstatus.get();
-
         let mut sstatus = LocalRegisterCopy::<u64, sstatus::Register>::new(0);
         sstatus.modify(sstatus::spp::Supervisor);
         sstatus.modify(sstatus::fs::Initial);
         state.guest_regs.sstatus = sstatus.get();
 
-        let mut hcounteren = LocalRegisterCopy::<u64, hcounteren::Register>::new(0);
-        hcounteren.modify(hcounteren::cycle.val(1));
-        hcounteren.modify(hcounteren::time.val(1));
-        hcounteren.modify(hcounteren::instret.val(1));
+        // Trap on guest acccess to all counters
```

Comment:
If we allow direct access, won't it require writing guest vCPU values to all HW counters when swapping the guest? Regarding FW counters, wasn't sure how they should be virtualized. Perhaps, we should snapshot the counters when swapping out the guest and re-configure with the stored initial count? @atishp04 Do you have any thoughts?

Reply:
> If we allow direct access, won't it require writing guest vCPU values to all HW counters when swapping the guest?

You can still trap accesses to counters which the guest hasn't configured. For the counters which have been configured by the guest, the trap handler just reads the same CSR the guest instruction was going to read. No point in taking a VM exit for that.

> Regarding FW counters, wasn't sure how they should be virtualized. Perhaps, we should snapshot the counters when swapping out the guest and re-configure with the stored initial count? 

Probably, though you need only save/restore the counters that have been enabled.

Reply:
mcountinhibit is M-mode only. Guests can't access that

Reply:
With supervisor counter delegation proposal, we will have scountinhibit. TSM can access that directly. If you use pass through, guests can access that as well. If you follow trap n emulate, you need to trap for that.

Reply:
Is there any way to tell from HS mode if the CPU supports `mcountinhibit`? In other words, is there a way for Salus to know that `sbi_pmu_counter_stop()` actually stops the counter from incrementing? From my read of OpenSBI, `sbi_pmu_counter_stop()` is a no-op on CPUs without `mcountinhibit` (unless `SBI_PMU_STOP_FLAG_RESET` is set).

Reply:
Not right now. The original implementation didn't enable SBI PMU extension without mcountinhibit support because you can't do much without the mcountinhibit. However, some of the platform vendors requested to allow it so that perf can be run on their platform for free running counters.

Once we have scountinhibit support, HS mode will be able figure out that.